File: driver/windows/log.cxx
#include <windows.h>

#include "log.h"
#include "util.h"

#include "_cgo_export.h"

void Log_Init() {}
#if defined(DEBUG)
void GoLogOut(enum EXCITON_LOG_LEVEL lvl, const char *str) {
  goDebugLog(lvl, (char *)str);
}
void GoLogOutW(enum EXCITON_LOG_LEVEL lvl, const WCHAR *str) {
  GoLogOut(lvl, exciton::util::ToUTF8String(str).c_str());
}
#else
void GoLogOut(enum EXCITON_LOG_LEVEL lvl, const char *str) {
  GoLogOutW(lvl, exciton::util::ToUTF16String(std::string(str)).c_str());
}
void GoLogOutW(enum EXCITON_LOG_LEVEL lvl, const WCHAR *str) {
  std::wstring wstr;
  switch (lvl) {
  case EXCITON_LOG_DEBUG:
    wstr = L"[DEBUG]\t";
    break;
  case EXCITON_LOG_INFO:
    wstr = L"[INFO]\t";
    break;
  case EXCITON_LOG_WARNING:
    wstr = L"[WARNING]\t";
    break;
  case EXCITON_LOG_ERROR:
    wstr = L"[ERROR]\t";
    break;
  }
  wstr += str;
  ::OutputDebugStringW(wstr.c_str());
}
#endif

File: driver/mac/log.m
#include "log.h"
#include "_cgo_export.h"

#if defined(DEBUG)
void Log_Init() {}
void GoLogOut(enum EXCITON_LOG_LEVEL lvl, NSString *str) {
  goDebugLog(lvl, (char *)[str UTF8String]);
}
#else
#import <os/log.h>
static os_log_t os_log_;
void Log_Init() {
  NSString *bundleIdentifier = [[NSBundle mainBundle] bundleIdentifier];
  if (bundleIdentifier) {
    os_log_ = os_log_create([bundleIdentifier UTF8String], "exciton_logging");
  } else {
    os_log_ = OS_LOG_DEFAULT;
  }
}
void GoLogOut(enum EXCITON_LOG_LEVEL lvl, NSString *str) {
  const char* s = [str UTF8String];
  GoLogOutC(lvl, s);
}
void GoLogOutC(enum EXCITON_LOG_LEVEL lvl, const char* s) {
  if (!s) s = "NULL";
  switch (lvl) {
  case EXCITON_LOG_ERROR:
    os_log_error(os_log_, "%{public}s", s);
    break;
  case EXCITON_LOG_INFO:
    os_log_info(os_log_, "%{public}s", s);
    break;
  case EXCITON_LOG_DEBUG:
    os_log_debug(os_log_, "%{public}s", s);
    break;
  case EXCITON_LOG_WARNING:
    os_log_error(os_log_, "%{public}s", s);
    break;
  }
}
#endif
